docs/rfcs/729-redesign-http-signer.md

```diff
@@ -0,0 +1,169 @@
+- Author: JinnyYi <github.com/JinnyYi>
+- Start Date: 2021-08-27
+- RFC PR: [beyondstorage/go-storage#729](https://github.com/beyondstorage/go-storage/issues/729)
+- Tracking Issue: [beyondstorage/go-storage#0](https://github.com/beyondstorage/go-storage/issues/0)
+
+# GSP-729: Redesign HTTP Signer
+
+- Updates:
+  - [GSP-706]: Deprecate `QuerySignHTTP()`.
+
+Previous discussion:
+- https://github.com/beyondstorage/go-storage/issues/726
+- https://matrix.to/#/#beyondstorage@gsp-706:matrix.org
+
+## Background
+
+In [GSP-706], we introduced `HTTPSigner` interface to support HTTP authenticating requests, which contains `QuerySignHTTP()`. `QuerySignHTTP()` is used to authenticate requests by using query parameters.
+
+During the implementation for services, we found the following problems:
+- There's no appropriate way to pass in some parameters for some operations like [How to pass partIndex into QuerySignHTTP for WriteMultipart](https://github.com/beyondstorage/go-storage/issues/726).
+- Supporting all the authenticating request operations in one function makes it lengthy and hard to maintain, especially for the services that support query string authentication like s3, gcs, etc.
+
+## Proposal
+
+I propose to split `QuerySignHTTP()` into multiple operations according to `op`:
+
+```go
+type HTTPSigner interface {
+    QuerySignHTTPRead(path string, expire time.Duration, ps ...types.Pair) (signedReq *http.Request, err error)
+    QuerySignHTTPWrite(path string, size int64, expire time.Duration, ps ...types.Pair) (signedReq *http.Request, err error)
+    // ...
+}
+```
+
+- `Read` and `Write` (add `Multiparter` related operations if needed) are the supported signature operations for now. Other operations SHOULD be introduced by new GSP.

```

Comment:
So we will only a `HTTPSigner` instead of `MultipartHTTPSinger`, `AppendHTTPSigner` and so on?

Reply:
Maybe we should group them by interfaces? And for `StoragerHTTPSigner`, I'm not sure whether we need to add operations for all the non-local functions or only for `read/write` temporarily (as some services only support add signatures to URLs that are contained in PUT and GET requests, or only for GET)?

Reply:
I prefer `StorageHTTPSigner`.

How about adding them one by one, and leave this question for the smarted us in the further to decide.

`Read/Write` is enough so far.